### PR TITLE
cscs-ci

### PR DESCRIPTION
**Pull request by boeschf**
# CI on CSCS machines

Leverages CSCS's CI-Ext facilites to run pipelines on some of the CSCS machines. As of now, there are 3 CI pipelines enabled:

| Pipeline name  | Description |
| ------------- | ------------- |
| `daint-mc` | Piz Daint MC partition: dual socket Intel Xeon E5-2695 v4 (2x18 cores Broadwell) |
| `daint-gpu` | Piz Daint Multicore partition: single socket Intel Xeon E5-2690 v3  (12 cores Haswell) + 1 NVidia P100 |
| `hohgant-cpu` | Hohgant Alps vCluster CPU partition: dual socket AMD EPYC 7742 (2x64 cores Zen2) |

The CI gets triggered when one submits a comment stating `cscs-ci run PIPELINE_NAME`, or if a whitelisted user is pushing a commit to a PR. At the moment, only PRs targeting the master branch will run this CI. The workflow behind the scenes is roughly as follows:

1. Github sends a webhook to cicd-ext-mw.cscs.ch (CI middleware)
2. CI middleware fetches your repository from github and pushes a mirror to gitlab
3. Gitlab sees a change in the repository and starts a pipeline (i.e. it uses the CI yaml as entrypoint)
4. The specified runner will take its input Dockerfile and execute docker build -f $DOCKERFILE ., where the build context is the whole (recursively) cloned repository

All pipelines run both on a single node and on two distinct nodes. Unit tests and examples (including Python examples) are executed.

### GPU fixes

- Pulled in some of the fixes from #2143 
- Fixed the AMD hipcc path by updating the compiler flags and properly marking the cuda files in CMake
- Fixed some Python tests (default selection of GPU)
- Fixed C++ example (proper compiler flags)
- Fixed some warnings on clang

### Incidental Changes

- Redesigned the `run_cpp_examples.sh` script to accept more arguments - this is useful for some of the pipelines where it is not possible to run several MPI examples back-to-back in the same workflow (likely due to an MPI bug).
- Python install prefix: convert to absolute path (was allowed to be relative path when specified through `ccmake` which led to wrong installation directory structure)

## Future Work

- add pipeline for Hohgant amdgpu partition: single socket AMD EPYC 7A53 (64 cores Zen3) + 8xMI200
- add pipeline for Hohgant nvgpu partition: single socket AMD EPYC 7713 (64 cores Zen3) + 4xA100

These pipelines require some work on our sides (develop and deploy appropriate gitlab runners, add default recipes etc) and, thus, I'll be addressing those in another PR.

## Current Limitations

The Hohgant runners are currently not picking up any pipelines, because of maintenance work on CSCS systems. Thus, I disabled the pipeline temporarily, and will enable it again end of August. During this maintenance period the runners targeting Piz Daint may be queued for longer times, as well.

### CODE REVIEW

**Comment by boeschf** [comment]
cscs-ci run

**Comment by boeschf** [comment]
cscs-ci run

**Comment by boeschf** [comment]
cscs-ci run

**Comment by boeschf** [comment]
cscs-ci run

**Comment by boeschf** [comment]
cscs-ci run daint-gpu

**Comment by boeschf** [comment]
cscs-ci run daint-mc

**Comment by boeschf** [comment]
cscs-ci run hohgant-cpu

**Comment by boeschf** [comment]
> cscs-ci run

**Comment by boeschf** [comment]
cscs-ci run daint-gpu

**Comment by boeschf** [comment]
cscs-ci run hohgant-cpu

**Comment by brenthuisman** [comment]
@boeschf: have you found a better solution for Pugi? Just spoke with @thorstenhater about dependencies, see also https://github.com/arbor-sim/arbor/pull/2153#issuecomment-1641810452, and we could for all submodules that have header-only versions, 'disable' using externally supplied versions, so, always pull in the submodule for those.

**Comment by boeschf** [comment]
Note that due to maintenance work the hohgant runners are currently not picking up jobs. This should be fixed by 31. August at the latest. There is a possibility that they'll work again by 16. August. The daint runners (including the gpu runner) are working, though. I don't think this should be a showstopper, though. I will disable the hohgant runner webhook for now, and re-enable end of month.

File: CMakeLists.txt [hunk 1]
```diff
@@ -418,7 +418,7 @@ if(ARB_WITH_GPU)
         target_compile_options(arbor-private-deps INTERFACE $<$<COMPILE_LANGUAGE:CXX>:${clang_options_}>)
         target_compile_options(arborenv-private-deps INTERFACE $<$<COMPILE_LANGUAGE:CXX>:${clang_options_}>)
     elseif(ARB_WITH_HIP_CLANG)
-        set(clang_options_ -DARB_HIP -xhip --amdgpu-target=gfx906 --amdgpu-target=gfx900)
+        set(clang_options_ -DARB_HIP -xhip --offload-arch=gfx906 --offload-arch=gfx900)
```
Comment (thorstenhater) [inline CMakeLists.txt:421]: This has been a þ in my side for quite some time, but never came round to fix it. Can we detect/generalise or consume user args, similar to `CMAKE_CUDA_ARCHITECTURES`? If not it's not a big deal.

File: arbor/util/pimpl_src.hpp [hunk 2]
```diff
@@ -14,7 +14,7 @@ template<typename T>
 pimpl<T>::~pimpl() = default;
 
 template<typename T>
-pimpl<T>::pimpl() noexcept = default;
+pimpl<T>::pimpl() noexcept {}
```
Comment (thorstenhater) [inline arbor/util/pimpl_src.hpp:17]: The usual wisdom is to use `default` over `{}`. What motivated this change?

File: ci/cscs/daint_gpu/Dockerfile.base [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+COPY --from=builder /etc/profile.d/z10_spack_environment.sh /etc/profile.d/z10_spack_environment.sh
+
+# Some boilerplate to get all paths correctly - fix_spack_install is part of the base image
+# and makes sure that all important things are being correctly setup
+RUN fix_spack_install
+
+# Finally install software that is needed, e.g. compilers
+# It is also possible to build compilers via spack and let all dependencies be handled by spack
+RUN apt-get -yqq update && apt-get -yqq upgrade \
+ && apt-get -yqq install build-essential gfortran \
```
Comment (thorstenhater) [inline ci/cscs/daint_gpu/Dockerfile.base:48]: `gfortran`? For `numpy`?!

File: ci/cscs/daint_gpu/pipeline.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
+include:
+  - remote: 'https://gitlab.com/cscs-ci/recipes/-/raw/master/templates/v2/.ci-ext.yml'
+
+stages:
+  - build_base # build stage is running on Kubernetes cluster
+  - build      # build stage is running on Kubernetes cluster
+  - test       # test stage is runnung on slurm cluster
```
Comment (thorstenhater) [inline ci/cscs/daint_gpu/pipeline.yml:7]:
```suggestion
  - test       # test stage is running on slurm cluster
```

File: ci/cscs/daint_mc/pipeline.yml [hunk 5]
```diff
@@ -0,0 +1,66 @@
+include:
+  - remote: 'https://gitlab.com/cscs-ci/recipes/-/raw/master/templates/v2/.ci-ext.yml'
+
+stages:
+  - build_base # build stage is running on Kubernetes cluster
+  - build      # build stage is running on Kubernetes cluster
+  - test       # test stage is runnung on slurm cluster
```
Comment (thorstenhater) [inline ci/cscs/daint_mc/pipeline.yml:7]:
```suggestion
  - test       # test stage is running on slurm cluster
```

File: ci/cscs/hohgant_cpu/pipeline.yml [hunk 6]
```diff
@@ -0,0 +1,141 @@
+include:
+  - remote: 'https://gitlab.com/cscs-ci/recipes/-/raw/master/templates/v2/.ci-ext.yml'
+
+stages:
+  - build_base # build stage is running on Kubernetes cluster
+  - build      # build stage is running on Kubernetes cluster
+  - test       # test stage is runnung on slurm cluster
```
Comment (thorstenhater) [inline ci/cscs/hohgant_cpu/pipeline.yml:7]:
```suggestion
  - test       # test stage is running on slurm cluster
```

File: ci/cscs/hohgant_cpu/pipeline.yml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+test-single-node:
+  extends: .container-runner-hohgant-zen2
+  stage: test
+  image: $PERSIST_IMAGE_NAME
+  script:
+    - cd /arbor.src
+    - build/bin/unit-modcc
+    - build/bin/unit-local
+    - build/bin/unit
+    #- scripts/run_cpp_examples.sh
```
Comment (thorstenhater) [inline ci/cscs/hohgant_cpu/pipeline.yml:37]: ?

File: python/CMakeLists.txt [hunk 8]
```diff
@@ -97,9 +97,13 @@ file(WRITE "${CMAKE_BINARY_DIR}/install-prefix" "${CMAKE_INSTALL_PREFIX}")
 execute_process(
     COMMAND ${PYTHON_EXECUTABLE} "${PROJECT_SOURCE_DIR}/scripts/where.py"
     INPUT_FILE "${CMAKE_BINARY_DIR}/install-prefix"
-    OUTPUT_VARIABLE ARB_PYTHON_LIB_PATH_DEFAULT
+    OUTPUT_VARIABLE ARB_PYTHON_LIB_PATH_DEFAULT_REL
     OUTPUT_STRIP_TRAILING_WHITESPACE)
 
+# convert to absolute path if needed (could be a relative path if ccmake was used)
```
Comment (thorstenhater) [inline python/CMakeLists.txt:103]: 👍

File: scripts/run_cpp_examples.sh
Comment (thorstenhater) [inline scripts/run_cpp_examples.sh:1]: Note for another time .... this is becoming a bit too large for a shell script
for my tastes. Suggestions? Python?

**Review by thorstenhater — CHANGES_REQUESTED** [review]
Hey,

looking good, some minor comments/questions.

File: ci/cscs/daint_gpu/Dockerfile.base (hunk 3 above)
Comment (boeschf) [reply ci/cscs/daint_gpu/Dockerfile.base:48]: oops, copied over from a template, gfortran obviously not needed

File: arbor/util/pimpl_src.hpp (hunk 2 above)
Comment (boeschf) [reply arbor/util/pimpl_src.hpp:17]: hipcc didn't like it if I recall correctly

Comment (thorstenhater) [reply arbor/util/pimpl_src.hpp:17]: Shame on it then. Ok, but if you have the bandwidth to spare, could you annotate accordingly?

Comment (boeschf) [reply arbor/util/pimpl_src.hpp:18]: done

File: ci/cscs/hohgant_cpu/pipeline.yml (hunk 7 above)
Comment (boeschf) [reply ci/cscs/hohgant_cpu/pipeline.yml]: removed comments

File: scripts/run_cpp_examples.sh
Comment (boeschf) [reply scripts/run_cpp_examples.sh:1]: yes, I agree it's getting lengthy. Python is the obvious choice - will translate the next time we need to change it

File: CMakeLists.txt
Comment (thorstenhater) [inline CMakeLists.txt:1]: Nice, although it'll conflict with the upgrade in checkpoints

**Review by thorstenhater — DISMISSED** [review]

**Comment by boeschf** [comment]
cscs-ci run daint-gpu

**Comment by boeschf** [comment]
cscs-ci run daint-mc

**Review by brenthuisman — APPROVED** [review]
Based on the fact that tests passed, including the cscs ones, and that nothing has changed but merge master after @thorstenhater's positive review, I'll merge.

**Pull request merged**